pkg/build/controller/buildpod/controller.go

```diff
@@ -0,0 +1,401 @@
+	kcoreclient "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset/typed/core/internalversion"
+	kcontroller "k8s.io/kubernetes/pkg/controller"
+	utilruntime "k8s.io/kubernetes/pkg/util/runtime"
+	"k8s.io/kubernetes/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/util/workqueue"
+
+	buildapi "github.com/openshift/origin/pkg/build/api"
+	buildclient "github.com/openshift/origin/pkg/build/client"
+	"github.com/openshift/origin/pkg/build/controller/common"
+	"github.com/openshift/origin/pkg/build/controller/policy"
+	strategy "github.com/openshift/origin/pkg/build/controller/strategy"
+	buildutil "github.com/openshift/origin/pkg/build/util"
+	osclient "github.com/openshift/origin/pkg/client"
+	oscache "github.com/openshift/origin/pkg/client/cache"
+)
+
+const (
+	// We must avoid processing build pods until the build and pod stores have synced.
+	// If they haven't synced, to avoid a hot loop, we'll wait this long between checks.
+	storeSyncedPollPeriod = 100 * time.Millisecond
+	maxRetries            = 60
+)
+
+// BuildPodController watches pods running builds and manages the build state
+type BuildPodController struct {
+	buildUpdater buildclient.BuildUpdater
+	secretClient kcoreclient.SecretsGetter
+
+	queue workqueue.RateLimitingInterface
+
+	buildStore oscache.StoreToBuildLister
+	podStore   cache.StoreToPodLister
+
+	buildStoreSynced func() bool
+	podStoreSynced   func() bool
+
+	runPolicies []policy.RunPolicy
+}
+
+// NewBuildPodController creates a new BuildPodController.
+func NewBuildPodController(buildInformer, podInformer cache.SharedIndexInformer, kc kclientset.Interface, oc osclient.Interface) *BuildPodController {
+	buildListerUpdater := buildclient.NewOSClientBuildClient(oc)
+	c := &BuildPodController{
+		buildUpdater: buildListerUpdater,
+		secretClient: kc.Core(), // TODO: Replace with cache client
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+	}
+
+	c.runPolicies = policy.GetAllRunPolicies(buildListerUpdater, buildListerUpdater)
+
+	c.buildStore.Indexer = buildInformer.GetIndexer()
+	c.podStore.Indexer = podInformer.GetIndexer()
+	podInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: c.updatePod,
+		DeleteFunc: c.deletePod,
+	})
+	buildInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addBuild,
+		UpdateFunc: c.updateBuild,
+	})
+
+	c.buildStoreSynced = buildInformer.HasSynced
+	c.podStoreSynced = podInformer.HasSynced
+
+	return c
+}
+
+// Run begins watching and syncing.
+func (c *BuildPodController) Run(stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	// Wait for the build store to sync before starting any work in this controller.
+	if !cache.WaitForCacheSync(stopCh, c.buildStoreSynced, c.podStoreSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	go wait.Until(c.worker, time.Second, stopCh)
```

Comment:
The workqueue ensures that only one key with the value "X" is processed at a time.